### PR TITLE
Error when repeatedly checking out a forked PR

### LINKED ISSUE #11238: Pull request from forked repo: "A branch with that name already exists."
### Describe the bug

After upgrading to 2.6.1 (apparently), re-selecting a PR branch from a forked repo fails with the error "A branch with that name already exists."

### Version & OS

- GH Desktop: 2.6.1
- MacOS: 10.14.6

### Steps to reproduce the behavior

1. Clone the storybook repo: https://github.com/storybookjs/storybook
2. Select a PR from a forked branch, e.g. `pr/13142`
3. Select a non-forked branch, e.g. `next`
4. Re-select the PR from the forked branch in #2
5. See error "A branch with that name already exists."

Note that selecting a second PR from the same fork, e.g. `pr/13429`, will succeed after performing these steps. And then selecting another branch and re-selecting `pr/13429` will fail with the same error.

### Expected behavior

I expect to be able to freely switch between branches, including PRs from forked repos.

### Actual behavior

Selecting a branch on a forked repo fails the second time around.

### Logs

```
2020-12-16T02:11:24.810Z - info: [ui] Executing fetch: git -c credential.helper= -c protocol.version=2 fetch --progress --prune github-desktop-BarryThePenguin (took 2.222s)
2020-12-16T02:11:25.178Z - error: [ui] `git branch pr/13429 github-desktop-BarryThePenguin/update-ember-cli` exited with an unexpected code: 128.
stderr:
fatal: A branch named 'pr/13429' already exists.

(The error was parsed as 26: A branch with that name already exists.)
```

### Additional context

- Checking out PRs from forked repos is a regular part of my workflow, and the problem seems to have been introduced very recently, probably in `v2.6.1`.
- Manually deleting the branch locally on the command line `git branch -d pr/13429` is a temporary workaround.

### PR DESCRIPTION
**Pull request by niik**
Closes #11238

## Description

This was an inadvertent side-effect of #10957. In [this refactor commit](https://github.com/desktop/desktop/commit/2e70798475) I inlined a call to `getLocalBranch` and  forgot to change the branch variable to use the value that the previously inlined function was called with. The result being that Desktop got confused and thought that there wasn't any existing branch for the PR and attempted to create it anyway.

This logic has been on my list to improve for a long time. The logic is complex and not very well documented so I took the liberty of cleaning things up and making them more robust. As part of this change I've updated the logic for finding local branches for a PR to not rely merely on the `pr/number` ref name format but rather look for local branches that are already tracking the specific remote.

I've also changed the way we look up matching PRs from relying solely on a comparing remote urls to the PR repo CloneURL to instead use `urlMatchesRemote` which is protocol agnostic and able to see that `git@github.com:desktop/desktop.git` is actually the same thing as both `https://github.com/desktop/desktop` and `https://github.com/desktop/desktop.git`

## Release notes

Notes: [Fixed] Checking out a pull request from a fork repeatedly now detects that there's already a branch that can be reused.

### CODE REVIEW

File: app/src/lib/stores/app-store.ts [hunk 1]
```diff
@@ -5294,119 +5295,74 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
-      )
-
-      log.error(error.message)
-      return this.emitError(error)
-    }
+    // Find an existing remote (regardless if set up by us or outside of
+    // Desktop). If we can't find one we'll create a Desktop for remote.
+    const remote =
+      remotes.find(r => urlMatchesRemote(headCloneUrl, r)) ??
+      (await addRemote(repository, forkRemoteName, headCloneUrl))
```
Comment (sergiou87) [inline app/src/lib/stores/app-store.ts]: This is something you figured out, so I won't take any credit for it 😂 , but I thought it was worth documenting it here maybe?

So as you explained to me, this `addRemote` could fail (silently as an "unhandled promise rejection" instead of crashing) if the repo already had a remote set up with this `forkRemoteName` name, but a different URL. For example, in the weird scenario where someone used Desktop (without the `ownerLogin` vs `headRepoOwner` changes in this PR) like this:

1. Let's say user `A` submits a PR to the repo `R`, using a branch that comes from the fork `FZ`, which belongs to user `Z`.
1. Then I use Desktop (prod, without this fixes) to checkout to that PR's branch, I'd get a remote called `github-desktop-A` pointing to the fork `FZ` (which belongs to user `Z`, not to user `A`).
1. Then I update Desktop to get the fixes from this PR.
1. After that I use Desktop again now to checkout to some other PR that `A` submitted to `R`, but now the source branch comes from the fork `FA` which actually belongs to `A`.
1. In that case, Desktop will look for a remote referencing `FA` using its clone URL, it will find nothing, so it will **try** to add, thanks to the fixes here, a remote like `github-desktop-A`, and it will fail because `github-desktop-A` was created previously to reference `FZ` instead 😢 

Of course, there is also the possibility of users tinkering manually with git remotes, which could end up in the same scenario but in a less confusing way 😂 

TLDR; should we catch and handle the error here as we discussed?

**Review by sergiou87 — DISMISSED** [review]
LGTM!

I'm probably biased because you explained the changes to me 😂 but reading the before and after versions, I definitely see an improvement in clarity and readability (aside from the actual fixes, of course).

Also nice catch with that `ownerLogin` vs `headRepoOwner` thing! 🤯 

The only thing worth considering is handling the `addRemote` error to make sure users are aware when that happens, otherwise I'd say :shipit:

File: app/src/lib/stores/app-store.ts (hunk 1 above)
Comment (niik) [reply app/src/lib/stores/app-store.ts]: :heart_eyes: This is so well documented. I don't believe my explanation was nearly as easy to follow. Thanks for taking the time to lay it out here

> TLDR; should we catch and handle the error here as we discussed?

Yeah I think it's worth it even though it's likely super rare. I'll get on it

File: app/src/lib/stores/app-store.ts [hunk 2]
```diff
@@ -5301,13 +5301,21 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+    // Desktop).
+    let remote = remotes.find(r => urlMatchesRemote(headCloneUrl, r))
+
+    // If we can't find one we'll create a Desktop fork remote.
+    if (remote === undefined) {
+      try {
+        const forkRemoteName = forkPullRequestRemoteName(headRepoOwner)
+        remote = await addRemote(repository, forkRemoteName, headCloneUrl)
+      } catch (e) {
+        this.emitError(e)
```
Comment (sergiou87) [inline app/src/lib/stores/app-store.ts]: Since (unless I'm missing something) this is a user-facing error, should we prefix the error with some text from us to give some context? Something like…
```suggestion
        this.emitError(`Couldn't checkout branch '${headRefName}', there was an error adding a remote for that branch: ${e.message}`)
```

**Review by sergiou87 — COMMENTED** [review]
Looks beautiful now (and it works 😂), thanks for the changes!! 😍 :shipit:

**Review by sergiou87 — APPROVED** [review]
